### PR TITLE
Bugfix/issue 153 token revalidation

### LINKED ISSUE #153: Implement refresh and access token system in backend
**Description:**
Develop a robust refresh and access token system in the backend to enhance application security.

**Tasks:**
- Modify authentication service to generate both access and refresh tokens
- Update login and registration processes
- Implement token refresh endpoint
- Update middleware for token validation
- Implement token revocation system

**Acceptance Criteria:**
- [ ] Modify the signAndSetCookie function to generate both access and refresh tokens
- [ ] Update the login, register, and OAuth callback functions to use the new token system
- [ ] Create a new endpoint for token refresh (/auth/refresh-token)
- [ ] Update the requireAuth middleware to validate access tokens
- [ ] Implement token revocation on logout
- [ ] Update error handling to return appropriate responses for token-related errors
- [ ] Implement rate limiting on token-related endpoints to prevent abuse
- [ ] Update the passport JWT strategy to work with the new token system
- [ ] Ensure all routes that require authentication use the new token system

**Additional Information:**
- Consider using JSON Web Tokens (JWT) for implementing the token system.
- Ensure that refresh tokens have a longer expiry time compared to access tokens.
- Implement proper encryption for storing refresh tokens if needed.
- Update the auth.service.ts file to include functions for handling the new token system.
- Modify the auth-routes.schema.ts to include validation for the new token endpoints.

**Related Dependencies or References:**

### PR DESCRIPTION
**Pull request by dsmabulage**
## Purpose
The purpose of this PR is to fix #153 

## Goals
Implement a token refresh flow to ensure continuous authentication by issuing new access tokens when the current ones expire.

## Approach
- Added logic to verify the access token and issue a new token using the refresh token when the access token is expired.
- Implemented a method to set both the new access token and refresh token as cookies in the response.
- Updated the authentication middleware to handle token expiration and automatically refresh tokens if a valid refresh token is provided.

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -1,24 +1,33 @@
...
+import { MenteeApplicationStatus, MentorApplicationStatus } from './enums'
+import { generateCertificate } from './services/admin/generateCertificate'
 
 export const signAndSetCookie = (res: Response, uuid: string): void => {
   const token = jwt.sign({ userId: uuid }, JWT_SECRET ?? '')
+  const refreshToken = jwt.sign({ userId: uuid }, REFRESH_JWT_SECRET ?? '', {
+    expiresIn: '10d'
+  })
 
   res.cookie('jwt', token, {
```
Comment (anjula-sack) [inline src/utils.ts:20]: This should be the access token

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/configs/linkedin-passport.ts [hunk 2]
```diff
@@ -68,7 +68,7 @@ passport.deserializeUser(async (primary_email: string, done) => {
 const cookieExtractor = (req: Request): string => {
   let token = null
   if (req?.cookies) {
-    token = req.cookies.jwt
+    token = req.cookies.accessToken
```
Comment (anjula-sack) [inline src/configs/linkedin-passport.ts:71]: We need a refresh token endpoint as well. Refer to this
https://medium.com/@techsuneel99/jwt-authentication-in-nodejs-refresh-jwt-with-cookie-based-token-37348ff685bf

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/configs/linkedin-passport.ts (hunk 2 above)
Comment (dileepainivossl) [reply src/configs/linkedin-passport.ts:71]: added the separate refresh token endpoint

Comment (dsmabulage) [reply src/configs/linkedin-passport.ts:71]: added a seperate endpoint for a refresh token

File: src/controllers/auth.controller.ts [hunk 3]
```diff
@@ -163,18 +164,33 @@ export const requireAuth = (
...
+        if (!refreshToken) {
+          return res.status(401).send('Access Denied. Please Login again.')
+        }
+
+        try {
+          const decoded = jwt.verify(refreshToken, REFRESH_JWT_SECRET) as {
+            userId: string
+          }
+
+          signAndSetCookie(res, decoded.userId)
```
Comment (anjula-sack) [inline src/controllers/auth.controller.ts:188]: Calling this updates both access and refresh token. But only the access token should be replaced

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/controllers/auth.controller.ts (hunk 3 above)
Comment (dsmabulage) [reply src/controllers/auth.controller.ts:188]: Does this logic need to be implemented the same as in the `/refresh` endpoint?

Comment (anjula-sack) [reply src/controllers/auth.controller.ts:188]: Here's a sample

```
app.post('/token', (req, res) => {
  const refreshToken = req.cookies['refreshToken'];

  if (!refreshToken || !refreshTokens.includes(refreshToken)) {
    return res.sendStatus(403);
  }

  jwt.verify(refreshToken, refreshTokenSecret, (err, user) => {
    if (err) {
      return res.sendStatus(403);
    }

    const newAccessToken = jwt.sign({ id: user.id, username: user.username }, accessTokenSecret, { expiresIn: '15m' });

    // Update accessToken cookie
    res.cookie('accessToken', newAccessToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: 'strict',
      maxAge: 15 * 60 * 1000 // 15 minutes
    });

    res.json({ message: 'Access token refreshed!' });
  });
});

```

Comment (dileepainivossl) [reply src/controllers/auth.controller.ts]: resolved

Comment (dsmabulage) [reply src/controllers/auth.controller.ts]: resolved

**Review by anjula-sack — CHANGES_REQUESTED** [review]
can you resolve the conflicts @dsmabulage

**Comment by dsmabulage** [comment]
> can you resolve the conflicts @dsmabulage

done

**Comment by anjula-sack** [comment]
After having a discussion regarding this. We decided not to implement this since we are using http only cookies already. Thank you for your contribution though.

**Pull request closed without merging**